### ISSUE TITLE
spaCy and tcmalloc compiled module leads to "Attempt to free invalid pointer"

### ISSUE THREAD

**Issue by f11r** [explosion/spaCy#938]
Running spacy after importing a module that is linked against `tcmalloc_minimal` leads to "Attempt to free invalid pointer".

# Steps to recreate
~~~
docker run -ti fedora:25 bash
dnf install python3-devel gperftools-devel gcc-c++ redhat-rpm-config
pip3 install spacy==1.7.3 cython==0.25.2
python3 -m spacy download en
cd /tmp
touch testclass.pyx
~~~
Create `setup.py` with the following content:
~~~
from setuptools import setup
from setuptools.extension import Extension


setup(
    name='testpackage',
    version='0.1',
    ext_modules=[
        Extension('testclass',
                  ['testclass.pyx'],
                  language='c++',
                  libraries=['tcmalloc_minimal']),
    ]
)
~~~
Run:
~~~
python3 setup.py install
python3 -c "import testclass; import spacy; spacy.load('en')('Test')"
~~~

Having the imports the other way around runs without problems.

## `gdb` dump
~~~
#0  0x00007ffff6c4992f in raise () from /lib64/libc.so.6
#1  0x00007ffff6c4b52a in abort () from /lib64/libc.so.6
#2  0x00007ffff67d6509 in tcmalloc::Log(tcmalloc::LogMode, char const*, int, tcmalloc::LogItem, tcmalloc::LogItem, tcmalloc::LogItem, tcmalloc::LogItem) () from /lib64/libtcmalloc_minimal.so.4
#3  0x00007ffff67d2319 in (anonymous namespace)::InvalidFree(void*) () from /lib64/libtcmalloc_minimal.so.4
#4  0x00007ffff67e0c2d in tc_free () from /lib64/libtcmalloc_minimal.so.4
#5  0x00007fffd56f8909 in __pyx_t_5spacy_6syntax_6_state_StateC::~__pyx_t_5spacy_6syntax_6_state_StateC (this=0x55555e19dff0, __in_chrg=<optimized out>) at spacy/syntax/parser.cpp:18569
#6  __pyx_f_5spacy_6syntax_6parser_6Parser_parseC (__pyx_v_self=0x7fffe6139d68, __pyx_v_tokens=0x5555615361c8, __pyx_v_length=1, __pyx_v_nr_feat=<optimized out>) at spacy/syntax/parser.cpp:9247
#7  0x00007fffd56e8b3b in __pyx_pf_5spacy_6syntax_6parser_6Parser_6__call__ (__pyx_v_tokens=<optimized out>, __pyx_v_self=0x7fffe6139d68) at spacy/syntax/parser.cpp:7654
#8  __pyx_pw_5spacy_6syntax_6parser_6Parser_7__call__ (__pyx_v_self=0x7fffe6139d68, __pyx_args=<optimized out>, __pyx_kwds=<optimized out>) at spacy/syntax/parser.cpp:7593
#9  0x00007ffff796e167 in PyObject_Call () from /lib64/libpython3.5m.so.1.0
#10 0x00007ffff7a2a5ea in PyEval_EvalFrameEx () from /lib64/libpython3.5m.so.1.0
#11 0x00007ffff7a2f9b3 in _PyEval_EvalCodeWithName () from /lib64/libpython3.5m.so.1.0
#12 0x00007ffff7a2fa93 in PyEval_EvalCodeEx () from /lib64/libpython3.5m.so.1.0
#13 0x00007ffff7999278 in function_call () from /lib64/libpython3.5m.so.1.0
#14 0x00007ffff796e167 in PyObject_Call () from /lib64/libpython3.5m.so.1.0
#15 0x00007ffff7984d04 in method_call () from /lib64/libpython3.5m.so.1.0
#16 0x00007ffff796e167 in PyObject_Call () from /lib64/libpython3.5m.so.1.0
#17 0x00007ffff79cd460 in slot_tp_call () from /lib64/libpython3.5m.so.1.0
#18 0x00007ffff796e167 in PyObject_Call () from /lib64/libpython3.5m.so.1.0
#19 0x00007ffff7a2a5ea in PyEval_EvalFrameEx () from /lib64/libpython3.5m.so.1.0
#20 0x00007ffff7a2f9b3 in _PyEval_EvalCodeWithName () from /lib64/libpython3.5m.so.1.0
#21 0x00007ffff7a2fa93 in PyEval_EvalCodeEx () from /lib64/libpython3.5m.so.1.0
#22 0x00007ffff7a2fabb in PyEval_EvalCode () from /lib64/libpython3.5m.so.1.0
#23 0x00007ffff7a4ed44 in run_mod () from /lib64/libpython3.5m.so.1.0
#24 0x00007ffff7a506e5 in PyRun_StringFlags () from /lib64/libpython3.5m.so.1.0
#25 0x00007ffff7a5074b in PyRun_SimpleStringFlags () from /lib64/libpython3.5m.so.1.0
#26 0x00007ffff7a67957 in Py_Main () from /lib64/libpython3.5m.so.1.0
#27 0x0000555555554b60 in main ()
~~~

**Comment by honnibal** [comment]
Thanks for the clear report. Do you have any insight into what's going on here? I don't call into that library directly, so I'm a bit puzzled.

**Comment by f11r** [comment]
One thing that's come to mind is that it is something similar/related to https://github.com/gperftools/gperftools/issues/110#issuecomment-133672248, i.e. the memory is allocated by one allocator and then free'd by another. Not sure why the switch would occur.

Also found https://github.com/tensorflow/tensorflow/issues/8146 and https://github.com/ros-perception/vision_opencv/issues/161 (same issue involving tensorflow and opencv), where the solution was to **not** `export LD_PRELOAD="/usr/lib/libtcmalloc_minimal.so.4" `.  
For our case, running `LD_PRELOAD="/usr/lib64/libtcmalloc_minimal.so.4"  python3 -c "import testclass; import spacy; spacy.load('en')('Test')"` **does** work, which seems to confirm that it is a "switching of allocators" problem.

Another interesting thing is that when I install spacy via conda the problem also does not occur (using this for my real project, haven't tested it for the minimal example posted above but I don't see any reason why it should be different for that).

# Bottom line

I don't think this is related to spacy, so closing this issue.

# Workaround

As a workaround I am currently using spacy installed via conda.
